core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy

```diff
@@ -115,23 +115,40 @@ class ConnectorConfiguratorTomcat {
     } else if (connectors.size() < 1) {
       return null
     } else {
-      throw new IllegalStateException("Unexpected server.xml format - one http connector expected at most")
+      throw new IllegalStateException("Unexpected server.xml format - one ajp connector expected at most")
     }
   }
 
   private boolean isSecured(Node connector) {
     return connector.@secure.toString().toLowerCase().trim() == "true"
   }
 
-  private void createNewConnector(Map<String, Object> attributes) {
-    server.Service.each { service ->
-      service.appendNode ("Connector", attributes)
+
+  private void createNewConnector(Node element) {
+    server.Service.each { Node service ->
+      service.appendNode(element, element.attributes(), element.value())
     }
   }
 
-  private void updateExistingConnector(Node Connector, Map<String, Object> attributes) {
-    attributes.each { attribute ->
-      Connector.@"${attribute.key}" = attribute.value
+  private void updateExistingConnector(Node connector, Node newConnector) {
+    // update attributes
+    newConnector.attributes().each { attribute ->
+      connector.@"${attribute.key}" = attribute.value
+    }
+
+    // update nodes
+    newConnector.each { Node newSubelement ->
+      if (newSubelement.name() == ConnectorUpgradeProtocolTomcat.ELEMENT_NAME) {
+        // create new element
+        connector.appendNode(newSubelement, newSubelement.attributes(), newSubelement.value())
```

Comment:
there should be a check whether `UpgradeProtocol` exists already, if not then execute this line else do the update or better:
````
if ( .... exists ....) {
  // update ...
} else {
  // create ...
} 
````
to make it as clear as possible for a 1st look.